File: 07-langs_particularities/1-php/src/Contexts/Posts/Application/Find/PostFinder.php
<?php
declare(strict_types=1);

namespace CodelyTv\Contexts\Posts\Application\Find;

use CodelyTv\Contexts\Posts\Domain\Post;
use CodelyTv\Contexts\Posts\Domain\PostDoesNotExistError;
use CodelyTv\Contexts\Posts\Domain\PostRepository;

final class PostFinder
{
	public function __construct(private readonly PostRepository $repository)
	{
	}

	/** @throws PostDoesNotExistError */
	public function find(string $id): Post
	{
		$post = $this->repository->search($id);

		if (null === $post) {
			throw new PostDoesNotExistError($id);
		}

		return $post;
	}
}

File: 07-langs_particularities/3-scala/src/main/scala/codely/contexts/rrss/posts/application/find/PostFinder.scala
package codely.contexts.rrss.posts.application.find

import codely.contexts.rrss.posts.domain.{Post, PostDoesNotExistError, PostId, PostRepository}

class PostFinder(repository: PostRepository) {
	def find(id: String): Either[PostDoesNotExistError, Post] = {
		repository.search(PostId(id)) match {
			case None => Left(PostDoesNotExistError(id))
			case Some(post) => Right(post)
		}
	}
}
